projects/09/Snake/Snake.jack

```diff
@@ -0,0 +1,161 @@
+		return;
+	}
+
+    /** Draws the snake on the screen. */
+    method void draw() {
+        do blockList.draw();
+        return;
+    }
+	
+	/** Moves the snake one step in the current direction. */
+	method void move() {
+		var BlockList lastBlockList;
+		var Block lastBlock, firstBlock;
+		
+		let lastBlockList = blockList.removeLast();
+		let lastBlock = lastBlockList.getBlock();
+		let firstBlock = blockList.getBlock();
+		
+		if (direction = 1) {
+			do lastBlock.setX(firstBlock.getX() + blockSize + 5);
+			do lastBlock.setY(firstBlock.getY());
+		}
+		if (direction = 2) {
+			do lastBlock.setX(firstBlock.getX() - blockSize - 5);
+			do lastBlock.setY(firstBlock.getY());
+		}
+		if (direction = 3) {
+			do lastBlock.setY(firstBlock.getY() - blockSize - 5);
+			do lastBlock.setX(firstBlock.getX());
+		}
+		if (direction = 4) {
+			do lastBlock.setY(firstBlock.getY() + blockSize + 5);
+			do lastBlock.setX(firstBlock.getX());
+		}
+		
+		do lastBlockList.appendBlockList(blockList); // move the last element to be the first;
+		let blockList = lastBlockList;
+		return;
+	}
+	
+	/** Sets the direction of the snake. U-Turns are not allowed. */
+	method void setDirection(int newDirection) {
+		if (((direction = 1) & (newDirection = 2)) | ((direction = 2) & (newDirection = 1))) {
+			// U-Turns are not allowed
+			return;
+		}
+		if (((direction = 3) & (newDirection = 4)) | ((direction = 4) & (newDirection = 3))) {
+			// U-Turns are not allowed
+			return;
+		}
+		let direction = newDirection;
+		return;
+	}
+	
+	/** Increases the size of the snake by 1 */
+	method void increaseSize() {
+		var Block newBlock, firstBlock;
+		var BlockList newBlockList;
+		
+		let firstBlock = blockList.getBlock();
+		// Set the coordinates to those of the first block so it will not be seen until the snake moves
+		let newBlock = Block.new(firstBlock.getX(), firstBlock.getY(), blockSize); 
+		let newBlockList = BlockList.new(newBlock);
+		
+		// Set the newly added block to be the last. The next time the snake will move it will become the first 
+		// and will be rendered if `draw()` is called.
+		do blockList.appendBlockList(newBlockList);
+		return;
+	}
+	
+	method int getHeadTop() {
+		var Block firstBlock;
+		let firstBlock = blockList.getBlock();
+		return firstBlock.getY();
+	}
+	
+	method int getHeadLeft() {
+		var Block firstBlock;
+		let firstBlock = blockList.getBlock();
+		return firstBlock.getX();
+	}
+	
+	method int getHeadRight() {
+		return getHeadLeft() + blockSize;
+	}
+	
+	method int getHeadBottom() {
+		return getHeadTop() + blockSize;
+	}
+	
+	/** Returns true if the snake's head collides with the given area */
+	method boolean doesCollide(int top, int bottom, in left, int right) {
+		var int snakeTop, snakeLeft, snakeRight, snakeBottom;
+		let snakeTop = getHeadTop();
+		let snakeLeft = getHeadLeft();
+		let snakeRight = getHeadRight();
+		let snakeBottom = getHeadBottom();
+		
+		if ((snakeTop > top) & (snakeTop < bottom)) {
```

Comment:
how can the head of the snake be both above the top limit and below the bottom limit at the same time?
am I missing something?

regardless, document the meaning of each if condition

Reply:
The top limit is actually lower than the bottom limit (The point [0,0] is located at the top left corner of the screen)